src/include/hash_ring.hpp

```diff
@@ -48,91 +48,101 @@ class HashRing : public ConsistentHashMap<ServerThread, H> {
     return unique_servers;
   }
 
+  bool insert_to_hash_ring(Address ip, unsigned tid) {
+    bool succeed;
+    for (unsigned virtual_num = 0; virtual_num < kVirtualThreadNum;
+         virtual_num++) {
+      succeed = HashRing::insert(ServerThread(ip, tid, virtual_num)).second;
+    }
+    return succeed;
+  }
+
+  void remove_from_hash_ring(Address ip, unsigned tid) {
+    for (unsigned virtual_num = 0; virtual_num < kVirtualThreadNum;
+         virtual_num++) {
+      HashRing::erase(ServerThread(ip, tid, virtual_num));
+    }
+  }
+
  private:
   std::unordered_set<ServerThread, ThreadHash> unique_servers;
 };
 
 typedef HashRing<GlobalHasher> GlobalHashRing;
 typedef HashRing<LocalHasher> LocalHashRing;
 
-template <typename H>
-bool insert_to_hash_ring(H& hash_ring, Address ip, unsigned tid) {
-  bool succeed;
-
-  for (unsigned virtual_num = 0; virtual_num < kVirtualThreadNum;
-       virtual_num++) {
-    succeed = hash_ring.insert(ServerThread(ip, tid, virtual_num)).second;
-  }
+class HashRingUtilInterface {
+ public:
+  virtual ServerThreadSet get_responsible_threads(
+      Address respond_address, const Key& key, bool metadata,
+      std::unordered_map<unsigned, GlobalHashRing>& global_hash_ring_map,
+      std::unordered_map<unsigned, LocalHashRing>& local_hash_ring_map,
+      std::unordered_map<Key, KeyInfo>& placement, SocketCache& pushers,
+      const std::vector<unsigned>& tier_ids, bool& succeed, unsigned& seed) = 0;
+
+  virtual ServerThreadSet responsible_global(
```

Comment:
If I remember correctly, we agreed that these two functions were not going to be called outside of this function and didn't need to be declared in the Interface, right?